vpc/service/allocate.go

```diff
@@ -412,3 +416,189 @@ func (vpcService *vpcService) ValidateAllocation(ctx context.Context, req *titus
 
 	return &titus.ValidationResponse{}, nil
 }
+
+type allocation struct {
+	id      string
+	region  string
+	account string
+	subnet  string
+}
+
+func (vpcService *vpcService) FixOldAllocations(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	tx, err := vpcService.db.BeginTx(ctx, &sql.TxOptions{
+		ReadOnly: true,
+	})
+	if err != nil {
+		return errors.Wrap(err, "Could not start database transaction")
+	}
+	defer func() {
+		_ = tx.Rollback()
+	}()
+
+	rows, err := tx.QueryContext(ctx,
+		`SELECT
+       ip_addresses.id, ip_addresses.region, subnets.account_id, ip_addresses.subnet_id
+FROM ip_addresses
+JOIN subnets ON ip_addresses.subnet_id = subnets.subnet_id
+WHERE ipv6address IS NULL`)
+	if err != nil {
+		return fmt.Errorf("Could not query IDs from ip_addresses: %w", err)
+	}
+
+	var allocations []allocation
+	for rows.Next() {
+		var alloc allocation
+		err = rows.Scan(&alloc.id, &alloc.region, &alloc.account, &alloc.subnet)
+		if err != nil {
+			return fmt.Errorf("Could not scan row: %w", err)
+		}
+
+		allocations = append(allocations, alloc)
+	}
+
+	err = tx.Commit()
+	if err != nil {
+		return fmt.Errorf("Cannot commit transaction to get outstanding, unconverted records: %w", err)
+	}
+
+	for _, alloc := range allocations {
+		err = vpcService.fixAllocation(ctx, alloc)
+		if err != nil {
+			return fmt.Errorf("Could not fix allocation %s: %w", alloc.id, err)
+		}
+	}
+
+	return nil
+}
+
+func (vpcService *vpcService) fixAllocation(ctx context.Context, alloc allocation) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	session, err := vpcService.ec2.GetSessionFromAccountAndRegion(ctx, ec2wrapper.Key{
+		AccountID: alloc.account,
+		Region:    alloc.region,
+	})
+	if err != nil {
+		return fmt.Errorf("Cannot get session: %w", err)
+	}
+
+	tx, err := vpcService.db.BeginTx(ctx, &sql.TxOptions{})
+	if err != nil {
+		return errors.Wrap(err, "Could not start database transaction")
+	}
+	defer func() {
+		_ = tx.Rollback()
+	}()
+
+	var ipv4addr string
+	var ipv6addr sql.NullString
+	row := tx.QueryRowContext(ctx,
+		"SELECT ip_address, ipv6address FROM ip_addresses WHERE id = $1 FOR NO KEY UPDATE",
+		alloc.id,
+	)
+	err = row.Scan(&ipv4addr, &ipv6addr)
+	if err != nil {
+		return fmt.Errorf("Cannot scan ip addresses: %w", err)
+	}
+	v4addr := net.ParseIP(ipv4addr)
+	v6addr := net.ParseIP(ipv6addr.String)
+	if !ipv6addr.Valid {
+		iface, err := vpcService.getStaticInterface(ctx, tx, session, alloc.subnet)
```

Comment:
Why don't you add the new code to allocate IP allocations using subnet CIDR reservations first? 

then you would be able to dynamically allocate an ENI on-demand and just use AWS's primitives for getting a V6 address.

Reply:
(See the suggestion above)